weblate/screenshots/forms.py

```diff
@@ -2,13 +2,78 @@
 #
 # SPDX-License-Identifier: GPL-3.0-or-later
 
+import io
+from typing import Any
+from urllib.parse import urlparse
+
+import requests
 from django import forms
+from django.conf import settings
+from django.core.files.uploadedfile import InMemoryUploadedFile
+from django.http.request import validate_host
 from django.template.loader import render_to_string
 from django.utils.html import format_html
+from django.utils.translation import gettext_lazy
 
 from weblate.screenshots.models import Screenshot
 from weblate.trans.forms import QueryField
 from weblate.utils.forms import SortedSelect
+from weblate.utils.validators import WeblateURLValidator
+
+
+class ScreenshotImageValidationMixin:
+    def clean_images(self, cleaned_data: dict[str, Any] | None) -> dict[str, Any]:
+        if cleaned_data is None:
+            cleaned_data = {}
+        image = cleaned_data.get("image")
+        image_url = cleaned_data.get("image_url")
+        if not image and not image_url:
+            raise forms.ValidationError(
+                gettext_lazy("You need to provide either image file or image URL.")
+            )
+
+        if image_url and not image:
+            cleaned_data["image"] = self.download_image(image_url)
+            image_field = Screenshot._meta.get_field("image")  # noqa: SLF001
+            image_field.run_validators(cleaned_data["image"])
+
+        cleaned_data.pop("image_url", None)
+        return cleaned_data
+
+    def download_image(self, url: str) -> InMemoryUploadedFile:
+        """Download image from the provided URL."""
+        if not validate_host(
+            urlparse(url).hostname or "", settings.ALLOWED_ASSET_DOMAINS
+        ):
+            raise forms.ValidationError(
+                gettext_lazy("Image URL domain is not allowed.")
+            )
+        try:
+            response = requests.get(url, timeout=2.0)
```

Comment:
Please use the wrapper from `weblate.utils.requests` instead (it sets `User-Agent`).